rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,547 @@
+	m_progress_dialog->show();
+
+	connect(m_progress_dialog, &QProgressDialog::canceled, reply_json, &QNetworkReply::abort);
+	connect(m_progress_dialog, &QProgressDialog::finished, m_progress_dialog, &QProgressDialog::deleteLater);
+
+	// clang-format off
+	connect(reply_json, &QNetworkReply::downloadProgress, [&](qint64 bytesReceived, qint64 bytesTotal)
+	{
+		m_progress_dialog->setMaximum(bytesTotal);
+		m_progress_dialog->setValue(bytesReceived);
+	});
+
+	connect(reply_json, &QNetworkReply::finished, [=]()
+	{
+		handle_reply(reply_json, &update_manager::handle_json, automatic, "Retrieved Json Info");
+	});
+	// clang-format on
+}
+
+bool update_manager::handle_reply(QNetworkReply* reply, std::function<bool(update_manager& man, const QByteArray&, bool)> func, bool automatic, const std::string& message)
+{
+	if (auto error = reply->error(); error != QNetworkReply::NoError)
+	{
+		reply->deleteLater();
+		if (error == QNetworkReply::OperationCanceledError)
+		{
+			QString error = reply->errorString();
+			LOG_ERROR(GENERAL, "[Auto-updater] Network Error: %s", error.toStdString());
+		}
+		return false;
+	}
+
+	// Read data from network reply
+	const QByteArray data = reply->readAll();
+	reply->deleteLater();
+
+	LOG_NOTICE(GENERAL, "[Auto-updater] %s", message);
+	if (!func(*this, data, automatic))
+	{
+		m_progress_dialog->close();
+
+		if (!automatic)
+			QMessageBox::warning(m_parent, tr("Auto-updater"), tr("An error occured during the auto-updating process.\nCheck the log for more information."));
+	}
+
+	return true;
+}
+
+bool update_manager::handle_json(const QByteArray& data, bool automatic)
+{
+	const QJsonObject json_data = QJsonDocument::fromJson(data).object();
+	int return_code             = json_data["return_code"].toInt(-255);
+
+	bool hash_found = true;
+
+	if (return_code < 0)
+	{
+		std::string error_message;
+		switch (return_code)
+		{
+		case -1: error_message = "Hash not found(Custom/PR build)"; break;
+		case -2: error_message = "Server Error - Maintenance Mode"; break;
+		case -3: error_message = "Server Error - Illegal Search"; break;
+		case -255: error_message = "Server Error - Return code not found"; break;
+		default: error_message = "Server Error - Unknown Error"; break;
+		}
+
+		if (return_code != -1)
+			LOG_ERROR(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+		else
+			LOG_WARNING(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+
+		// If a user clicks "Check for Updates" with a custom build ask him if he's sure he wants to update to latest version
+		if (!automatic && return_code == -1)
+		{
+			if (QMessageBox::question(m_progress_dialog, tr("Auto-updater"), tr("You're currently using a custom or PR build.\n\nDo you want to update to the latest official rpcs3 version?"),
+			        QMessageBox::Yes | QMessageBox::No) == QMessageBox::No)
+			{
+				m_progress_dialog->close();
+				return true;
+			}
+
+			hash_found = false;
```

Comment:
You should wrap this in the following two calls for the same reasons as before with the close stuff.
```
m_progress_dialog->hide();
// ...
// this code
// ...
m_progress_dialog->show();
```

Reply:
Before it made sense because the update was finished but leaving it open makes it clear that it's not finished, leaving as it is makes more sense imo.

Reply:
well it looks like crap if you abort and it glitches around but whatever